### PR TITLE
OutboundHandler should handle errors when sending responses

### PR DESCRIPTION
**Pull request by dnhatn**
The OutboundHandler should handle errors that occur during sending responses. Specifically, if sending a response encounters an exception, it should attempt to send an error response. If sending the error response fails, the OutboundHandler should close the channel and log a warning. This change would reduce the burden on callers, as they would no longer need to handle errors for these cases.

### CODE REVIEW

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I don't think we should relax this. Instead, the implementation should take steps to only complete its transport channel once at the very end of its execution.

**Comment by DaveCTurner** [comment]
In particular, if there's a way to call `onTaskFinished` a second time then the task wasn't really finished the first time it was called.

**Comment by dnhatn** [comment]
Thanks, David. I've understood your feedback and pushed https://github.com/elastic/elasticsearch/pull/105255/commits/b285f38dc4ce4c67d9225c5bdf9269a55d16b27d. However, I'm not comfortable with the new approach and slightly prefer the previous one because unregistering a task is idempotent. Any suggestion? Also, the previous approach is similar to what we have in TcpTransportChannel. If we decide to proceed with this, we might need to adjust TcpTransportChannel as well.

**Comment by DaveCTurner** [comment]
I don't think this is right either, we shouldn't be completing this listener twice. Can you give me a bit more context about the problem you're trying to solve here?

**Comment by dnhatn** [comment]
@DaveCTurner Sure, a CircuitBreakingException may occur when sending a large response using `sendResponse(TransportResponse)` in ESQL. This happens when we reserve bytes for the intermediate buffer used by OutputStream, such as RecyclerBytesStreamOutput. In such cases, both `sendResponse(TransportResponse)` and `sendResponse(Exception)` will call onTaskFinished twice, resulting in the test failing with `assertOnce` Releasable.

**Comment by DaveCTurner** [comment]
Hmm that's still not informative enough really. Can you add a test case showing the problem perhaps? Or share some stack traces showing how these things are being called? It still seems to me like a bug in the caller to invoke `sendResponse(Exception)` after invoking `sendResponse(TransportResponse)`. We shouldn't be doing that.

**Comment by DaveCTurner** [comment]
Oh wait I think I see, `OutboundHandler` doesn't look like it does the right thing with serialization exceptions from responses. If it hits an exception in `OutboundHandler#sendResponse` it should handle it by sending the exception down over the channel, not throwing it back out to the caller. Is that what you mean?

I guess it shouldn't be able to hit an exception in `OutboundHandler#sendErrorResponse`, but the only thing we can reasonably do in that situation is log an error and close the channel.

**Comment by dnhatn** [comment]
@DaveCTurner I think you're right. Here is an example:

```
[2024-02-07T19:11:41,960][WARN ][o.e.t.OutboundHandler    ] [test-cluster-1] failed to serialize outbound message [Response{907}{false}{false}{false}{class org.elasticsearch.compute.operator.exchange.ExchangeResponse}]
org.elasticsearch.common.breaker.CircuitBreakingException: [request] Data too large, data for [serializing page] would be [524570158/500.2mb], which is larger than the limit of [322122547/307.1mb]
	at org.elasticsearch.common.breaker.ChildMemoryCircuitBreaker.circuitBreak(ChildMemoryCircuitBreaker.java:91) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.common.breaker.ChildMemoryCircuitBreaker.limit(ChildMemoryCircuitBreaker.java:181) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.common.breaker.ChildMemoryCircuitBreaker.addEstimateBytesAndMaybeBreak(ChildMemoryCircuitBreaker.java:118) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.compute.operator.exchange.ExchangeResponse.writeTo(ExchangeResponse.java:53) ~[?:?]
	at org.elasticsearch.transport.OutboundMessage.serialize(OutboundMessage.java:71) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.transport.OutboundHandler.sendMessage(OutboundHandler.java:186) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.transport.OutboundHandler.sendResponse(OutboundHandler.java:140) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.transport.TcpTransportChannel.sendResponse(TcpTransportChannel.java:61) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:38) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.support.ChannelActionListener.lambda$onResponse$0(ChannelActionListener.java:31) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:356) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:31) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:19) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.ActionListener.respondAndRelease(ActionListener.java:289) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.compute.operator.exchange.ExchangeSinkHandler.notifyListeners(ExchangeSinkHandler.java:145) ~[?:?]
	at org.elasticsearch.compute.operator.exchange.ExchangeSinkHandler$LocalExchangeSink.addPage(ExchangeSinkHandler.java:60) ~[?:?]
	at org.elasticsearch.compute.operator.exchange.ExchangeSinkOperator.doAddInput(ExchangeSinkOperator.java:78) ~[?:?]
	at org.elasticsearch.compute.operator.SinkOperator.addInput(SinkOperator.java:32) ~[?:?]
	at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:214) ~[?:?]
	at org.elasticsearch.compute.operator.Driver.run(Driver.java:139) ~[?:?]
	at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:327) ~[?:?]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1583) ~[?:?]
[2024-02-07T19:11:41,964][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [test-cluster-1] fatal error in thread [elasticsearch[test-cluster-1][esql_worker][T#10]], exiting
java.lang.AssertionError: java.lang.Exception: wrapped[org.elasticsearch.transport.RequestHandlerRegistry$$Lambda/0x000000d0020fa660@5ee46246, org.elasticsearch.tasks.TaskManager$$Lambda/0x000000d00212b518@5de47026]
	at org.elasticsearch.core.Releasables$4.assertFirstRun(Releasables.java:153) ~[elasticsearch-core-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.core.Releasables$4.close(Releasables.java:158) ~[elasticsearch-core-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.transport.TaskTransportChannel.release(TaskTransportChannel.java:57) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:49) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:37) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:62) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:73) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:358) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:31) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:19) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.ActionListener.respondAndRelease(ActionListener.java:289) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.compute.operator.exchange.ExchangeSinkHandler.notifyListeners(ExchangeSinkHandler.java:145) ~[?:?]
	at org.elasticsearch.compute.operator.exchange.ExchangeSinkHandler$LocalExchangeSink.addPage(ExchangeSinkHandler.java:60) ~[?:?]
	at org.elasticsearch.compute.operator.exchange.ExchangeSinkOperator.doAddInput(ExchangeSinkOperator.java:78) ~[?:?]
	at org.elasticsearch.compute.operator.SinkOperator.addInput(SinkOperator.java:32) ~[?:?]
	at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:214) ~[?:?]
	at org.elasticsearch.compute.operator.Driver.run(Driver.java:139) ~[?:?]
	at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:327) ~[?:?]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.lang.Exception: wrapped[org.elasticsearch.transport.RequestHandlerRegistry$$Lambda/0x000000d0020fa660@5ee46246, org.elasticsearch.tasks.TaskManager$$Lambda/0x000000d00212b518@5de47026]
	at org.elasticsearch.core.Releasables$4.assertFirstRun(Releasables.java:152) ~[elasticsearch-core-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.core.Releasables$4.close(Releasables.java:158) ~[elasticsearch-core-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.transport.TaskTransportChannel.release(TaskTransportChannel.java:57) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:40) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.support.ChannelActionListener.lambda$onResponse$0(ChannelActionListener.java:31) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
	at org.elasticsearch.action.ActionListener.run(ActionListener.java:356) ~[elasticsearch-8.13.0-SNAPSHOT.jar:?]
```

**Comment by DaveCTurner** [comment]
Gotcha, yeah, we shouldn't be relying on `ChannelActionListener` to catch an exception and send it back down, this should all be handled in the `OutboundHandler`.

**Comment by dnhatn** [comment]
@DaveCTurner, are you okay with the changes made in the latest commits? I can add a test for that.

**Review by DaveCTurner — COMMENTED** [review]
Looks like the right sort of idea, although I think you're missing a `response.decRef()`. Also `sendErrorResponse` needs adjusting similarly and must close the channel on failure. And finally since these things can no longer throw, we can remove the `throws IOException` from a whole tree of callers.

**Comment by dnhatn** [comment]
> although I think you're missing a response.decRef()

I think we are good here since the `onAfter` calls this

@DaveCTurner Thank you for your feedback. Can you take another look?

File: server/src/test/java/org/elasticsearch/transport/OutboundHandlerTests.java [hunk 1]
```diff
@@ -317,6 +322,101 @@ public void onResponseSent(long requestId, String action, Exception error) {
...
+            sendListener.onFailure(new IOException("failed"));
+        }
+        assertEquals(requestId, requestIdRef.get());
+        assertEquals(action, actionRef.get());
+        assertEquals(response, responseRef.get());
+        assertThat(exceptionRef.get().getMessage(), equalTo("simulated"));
+        assertTrue(response.released.get());
+    }
+
+    public void testFailToSendHandshakeResponse() {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/transport/OutboundHandlerTests.java:377]: Well spotted, we do need different behaviour here indeed 👍

File: server/src/test/java/org/elasticsearch/transport/OutboundHandlerTests.java [hunk 2]
```diff
@@ -317,6 +322,101 @@ public void onResponseSent(long requestId, String action, Exception error) {
...
+                assertThat(requestIdRef.get(), equalTo(0L));
+                requestIdRef.set(requestId);
+                assertNull(actionRef.get());
+                actionRef.set(action);
+                assertNull(responseRef.get());
+                responseRef.set(response);
+            }
+
+            @Override
+            public void onResponseSent(long requestId, String action, Exception error) {
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/transport/OutboundHandlerTests.java:352]: Could we also assert that `channel.getMessageCaptor()` has seen an exception message here?

File: server/src/test/java/org/elasticsearch/transport/OutboundHandlerTests.java [hunk 3]
```diff
@@ -317,6 +322,101 @@ public void onResponseSent(long requestId, String action, Exception error) {
...
+                assertThat(requestIdRef.get(), equalTo(requestId));
+                assertThat(actionRef.get(), equalTo(action));
+                exceptionRef.set(error);
+            }
+        });
+        Compression.Scheme compress = randomFrom(compressionScheme, null);
+        try {
+            handler.sendResponse(version, channel, requestId, action, response, compress, false, ResponseStatsConsumer.NONE);
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/transport/OutboundHandlerTests.java:360]: Could we also have a test which fails to send both the response _and_ the resulting exception?

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java [hunk 4]
```diff
@@ -137,13 +139,31 @@ void sendResponse(
             compressionScheme
         );
         response.mustIncRef();
-        sendMessage(channel, message, responseStatsConsumer, () -> {
+        try {
+            sendMessage(channel, message, responseStatsConsumer, () -> {
+                try {
+                    messageListener.onResponseSent(requestId, action, response);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/OutboundHandler.java:145]: Ah I see, we call this on the exception paths too. It seems wrong to call `onResponseSent` if we didn't actually send the response, but that's how it works today. We can clean this up later.

**Review by DaveCTurner — COMMENTED** [review]
Looks good, just a few suggestions.

Could we also add some assertions into `ChannelActionListener` to say that the `channel.sendResponse` calls never throw now?

**Comment by dnhatn** [comment]
@DaveCTurner I think I have addressed all your feedback. Can you take another look? Thank you!

File: server/src/main/java/org/elasticsearch/action/support/ChannelActionListener.java [hunk 5]
```diff
@@ -28,16 +28,26 @@ public ChannelActionListener(TransportChannel channel) {
 
     @Override
     public void onResponse(Response response) {
-        ActionListener.run(this, l -> l.channel.sendResponse(response));
+        try {
+            channel.sendResponse(response);
+        } catch (RuntimeException e) {
+            final String message = format("channel [%s] threw exceptions on sendResponse", channel);
+            assert false : new AssertionError(message, e);
+            logger.warn(() -> message, e);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/ChannelActionListener.java:36]:
```suggestion
            logger.error(() -> message, e);
```

File: server/src/main/java/org/elasticsearch/action/support/ChannelActionListener.java [hunk 6]
```diff
@@ -28,16 +28,26 @@ public ChannelActionListener(TransportChannel channel) {
...
     public void onFailure(Exception e) {
         try {
             channel.sendResponse(e);
-        } catch (Exception sendException) {
+        } catch (RuntimeException sendException) {
             sendException.addSuppressed(e);
-            logger.warn(() -> format("Failed to send error response on channel [%s]", channel), sendException);
+            final String message = format("channel [%s] threw exceptions on sendResponse", channel);
+            assert false : new AssertionError(message, sendException);
+            logger.warn(() -> message, sendException);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/ChannelActionListener.java:49]:
```suggestion
            logger.error(() -> message, sendException);
```

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java [hunk 7]
```diff
@@ -137,13 +139,25 @@ void sendResponse(
...
+                } finally {
+                    response.decRef();
+                }
+            });
+        } catch (Exception ex) {
+            if (isHandshake) {
+                logger.warn(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/OutboundHandler.java:152]:
```suggestion
                logger.error(
```

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java [hunk 8]
```diff
@@ -156,11 +170,17 @@ void sendErrorResponse(
...
+    ) {
         TransportVersion version = TransportVersion.min(this.version, transportVersion);
         RemoteTransportException tx = new RemoteTransportException(nodeName, channel.getLocalAddress(), action, error);
         OutboundMessage.Response message = new OutboundMessage.Response(threadPool.getThreadContext(), tx, version, requestId, false, null);
-        sendMessage(channel, message, responseStatsConsumer, () -> messageListener.onResponseSent(requestId, action, error));
+        try {
+            sendMessage(channel, message, responseStatsConsumer, () -> messageListener.onResponseSent(requestId, action, error));
+        } catch (Exception sendException) {
+            sendException.addSuppressed(error);
+            logger.warn(() -> format("Failed to send error response on channel [%s], closing channel", channel), sendException);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/OutboundHandler.java:181]:
```suggestion
            logger.error(() -> format("Failed to send error response on channel [%s], closing channel", channel), sendException);
```

**Review by DaveCTurner — APPROVED** [review]
Looks good, I suggested changing some `warn` to `error` logs since these things should be impossible in practice, makes it a bit easier to spot them in logs. Otherwise LGTM

**Comment by dnhatn** [comment]
Thanks David!

**Pull request merged**